### PR TITLE
Update username regex and more

### PR DESCRIPTION
**Pull request by FrankApiyo**
### Changes introduced
- Use email address to set username when provided username is invalid
- Update default username regex (don't allow less than 3 characters)
- Update tests
- Update CI

### CODE REVIEW

File: tests/test_viewsets.py [hunk 1]
```diff
@@ -83,6 +85,77 @@ def test_returns_data_entry_template_on_missing_username_claim(self):
...
+                "family_name": "bob",
+                "given_name": "just bob",
+                "username": "bob@example.com",
+                "email": "bo@example.com",
+            }
+
+            data = {"id_token": "sadsdaio3209lkasdlkas0d.sdojdsiad.iosdadia"}
+            request = self.factory.post("/", data=data)
+            response = view(request, auth_server="default")
+            self.assertEqual(response.status_code, 400)
```
Comment (ukanga) [inline tests/test_viewsets.py:156]: Could we also add a check for the error message we expect?

File: tox.ini [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [tox]
 envlist =
-    py{38,39}-django{32,40}
+    py{38,39,312}-django{32,40}
```
Comment (ukanga) [inline tox.ini:3]: [Ona Data](https://github.com/onaio/onadata) tests on 3.10. Can we drop `38` and `39`? Can we add 3.10 and perhaps 3.11?

File: tests/test_viewsets.py (hunk 1 above)
Comment (FrankApiyo) [reply tests/test_viewsets.py:156]: done!

File: tox.ini (hunk 2 above)
Comment (FrankApiyo) [reply tox.ini]: done!

**Review by ukanga — APPROVED** [review]

**Pull request merged**